pyqtorch/utils.py

```diff
@@ -142,3 +143,33 @@ def density_mat(state: Tensor) -> Tensor:
     state = torch.permute(state, batch_first_perm).reshape(batch_size, 2**n_qubits)
     undo_perm = (1, 2, 0)
     return torch.permute(torch.einsum("bi,bj->bij", (state, state.conj())), undo_perm)
+
+
+def promote_ope(operator: Tensor, target: int, n_qubits: int) -> Tensor:
+    """
+    Promotes `operator` to the size of the circuit (number of qubits and batch).
+    Targeting the first qubit implies target = 0, so target > n_qubits - 1.
+
+    Args:
+        operator (Tensor): The operator tensor to be promoted.
+        target (int): The index of the target qubit to which the operator is applied.
+            Targeting the first qubit implies target = 0, so target > n_qubits - 1.
+        n_qubits (int): Number of qubits in the circuit.
+
+    Returns:
+        Tensor: The promoted operator tensor.
+
+    Raises:
+        ValueError: If `target` is outside the valid range of qubits.
+    """
+    if target > n_qubits - 1:
+        raise ValueError("The target must be a valid qubit index within the circuit's range.")
+
+    qubits = torch.arange(0, n_qubits)
+    for qubit in qubits:
+        if target > qubit:
```

Comment:
Can't you use [torch.where](https://pytorch.org/docs/stable/generated/torch.where.html) ?

Reply:
What do you think of this version ?

```python3
def promote_op(operator: Tensor, target: int, n_qubits: int) -> Tensor:
    """
    Promotes `operator` to the size of the circuit (number of qubits and batch).
    Targeting the first qubit implies target = 0, so target > n_qubits - 1.

    Args:
        operator (Tensor): The operator tensor to be promoted.
        target (int): The index of the target qubit to which the operator is applied.
            Targeting the first qubit implies target = 0, so target > n_qubits - 1.
        n_qubits (int): Number of qubits in the circuit.

    Returns:
        Tensor: The promoted operator tensor.

    Raises:
        ValueError: If `target` is outside the valid range of qubits.
    """
    if target > n_qubits - 1:
        raise ValueError("The target must be a valid qubit index within the circuit's range.")

    qubits = torch.arange(0, n_qubits)
    # Define I instead of importing it from pyqtorch.primitive to avoid circular import
    identity = torch.tensor([[[1.0 + 0.0j], [0.0 + 0.0j]], [[0.0 + 0.0j], [1.0 + 0.0j]]])
    for qubit in qubits:
        if target != qubit:
            operator = torch.where(
                target > qubit,torch.kron(identity, operator.contiguous()),torch.kron(operator.contiguous(), identity)
                )
    return operator

Reply:
What you can do to avoid circular import is a local one in the function:

```
def promote_op(...)
from pyqtorch.primitive import I
```

I would try to remove `target` while creating `qubits`. This would avoid checking for the difference.

Don't you need the case where `target < qubit` ?

Reply:
What do you think of this version ?

```python3
def promote_op(operator: Tensor, target: int, n_qubits: int) -> Tensor:
    from pyqtorch.primitive import I 
    """
    Promotes `operator` to the size of the circuit (number of qubits and batch).
    Targeting the first qubit implies target = 0, so target > n_qubits - 1.

    Args:
        operator (Tensor): The operator tensor to be promoted.
        target (int): The index of the target qubit to which the operator is applied.
            Targeting the first qubit implies target = 0, so target > n_qubits - 1.
        n_qubits (int): Number of qubits in the circuit.

    Returns:
        Tensor: The promoted operator tensor.

    Raises:
        ValueError: If `target` is outside the valid range of qubits.
    """
    if target > n_qubits - 1:
        raise ValueError("The target must be a valid qubit index within the circuit's range.")
    qubits = torch.arange(0, n_qubits)
    qubits = torch.cat([qubits[:target],qubits[target+1:]])
    for qubit in qubits:
        operator = torch.where(
            target > qubit,torch.kron(I(target).unitary(), operator.contiguous()),torch.kron(operator.contiguous(), I(target).unitary())
            )
    return operator
```

The case `target < qubit` is treat when the condition `target > qubit` in `torch.where()` is `False` as `torch.where(condition, input, other)` return a tensor of elements selected from either `input` or `other`, depending on `condition`.

Reply:
Feels like it. Isn't it possible to do:

```python
qubits = torch.cat([qubits[:target],qubits[target+1:]])
```

as

```python
qubits = qubits[qubits != target]
```